cql3/restrictions/statement_restrictions.hh

```diff
@@ -105,8 +105,30 @@ private:
     bool _has_queriable_regular_index = false, _has_queriable_pk_index = false, _has_queriable_ck_index = false;
 
     std::optional<expr::expression> _where; ///< The entire WHERE clause.
-    std::vector<expr::expression> _clustering_prefix_restrictions; ///< Parts of _where defining the clustering slice.
-    std::vector<expr::expression> _partition_range_restrictions; ///< Parts of _where defining the partition range.
+
+    /// Parts of _where defining the clustering slice.
+    ///
+    /// Meets all of the following conditions:
+    /// 1. all elements must be simultaneously satisfied (as restrictions) for _where to be satisfied
+    /// 2. each element is an atom or a conjunction of atoms
+    /// 3. either all atoms (across all elements) are multi-column or they are all single-column
+    /// 4. if single-column, then:
+    ///   4.1 all atoms from an element have the same LHS, which we call the element's LHS
+    ///   4.2 each element's LHS is different from any other element's LHS
+    ///   4.3 the list of each element's LHS, in order, forms a clustering-key prefix
+    ///   4.4 elements other than the last have only EQ or IN atoms
+    ///   4.5 the last element has only EQ, IN, or is_slice() atoms
+    /// 5. if multi-column, then each element is a binary_operator
+    std::vector<expr::expression> _clustering_prefix_restrictions;
```

Comment:
(unrelated: should be a single expression, using a conjunction if needed)

Reply:
I disagree.  Separating `c1>0 AND c1<=10` from `c2=1` makes clustering-range computation much simpler.

Reply:
A conjunction is a vector of expressions. So we'll be replacing the for loop iterating on the existing vector with the for loop iterating on the inner conjunction vector.

In fact, it probably works already.

```c++
value_set possible_lhs_values(const column_definition* cdef, const expression& expr, const query_options& options) {
    const auto type = cdef ? get_value_comparator(cdef) : long_type.get();
    return std::visit(overloaded_functor{
            [] (bool b) {
                return b ? unbounded_value_set : empty_value_set;
            },
            [&] (const conjunction& conj) {
                return boost::accumulate(conj.children, unbounded_value_set,
                        [&] (const value_set& acc, const expression& child) {
                            return intersection(
                                    std::move(acc), possible_lhs_values(cdef, child, options), type);
                        });
            },
```

You already have the inner loop, it's just never exercised.

Reply:
> A conjunction is a vector of expressions.

Yes.

> So we'll be replacing the for loop iterating on the existing vector with the for loop iterating on the inner conjunction vector.

Yes.

> In fact, it probably works already.

No.  The loop you quote and this loop are very different:
https://github.com/scylladb/scylla/blob/029991bfc236f0a3dfc1e8205b1930c85ae98b09/cql3/restrictions/statement_restrictions.cc#L1024

Reply:
Then one of them is buggy. They're defined in exactly the same way.

Reply:
Ah, I think the vector has some pre-processing in that it is sorted according to clustering key column order. Is that correct?

Reply:
Not sure I understand how `possible_lhs_values` and `get_single_column_clustering_bounds` are the same?  One gets a column's value_set, the other combines several of these value_sets into a vector.

Reply:
> Ah, I think the vector has some pre-processing in that it is sorted according to clustering key column order. Is that correct?

Yes; see 4.3 in the comment above.

Reply:
Ok, I understand it better now. The vector isn't just a collection of expressions, it has 1:1 correspondence with the clustering key elements (or a prefix of them). I knew that and forgot. I agree it totally shouldn't be a single expression.